### PR TITLE
Enhancement/default validation providers

### PR DESCRIPTION
**Pull request by lilHermit**
This allows plugin maintainers to load validation providers during bootstrap therefore reducing the traction to install.

### CODE REVIEW

File: src/Validation/Validator.php [hunk 1]
```diff
@@ -211,6 +219,43 @@ public function getProvider($name)
...
 
     /**
+     * Returns the default provider stored under that name if it exists.
+     *
+     * @param string $name The name under which the provider should be retrieved.
+     * @return object|string|null
+     */
+    public static function getDefaultProvider($name)
+    {
+        if (isset(self::$_defaultProviders[$name])) {
```
Comment (dereuromark) [inline src/Validation/Validator.php]: Usually you bail early (inverted):

    if (!isset) return null

**Comment by dereuromark** [comment]
Sounds reasonable.

File: tests/TestCase/Validation/ValidatorTest.php [hunk 2]
```diff
@@ -1879,4 +1879,36 @@ protected function assertProxyMethod($validator, $method, $extra = null, $pass =
...
+     * @return void
+     */
+    public function testGetDefaultProvider()
+    {
+        Validator::addDefaultProvider('test-provider', '\MyNameSpace\Validation\MyProvider');
+        $this->assertEquals(Validator::getDefaultProvider('test-provider'), '\MyNameSpace\Validation\MyProvider', 'Default provider `test-provider` is missing');
+
+        $this->assertNull(Validator::getDefaultProvider('invalid-provider'), 'Default provider (`invalid-provider`) should be missing');
+        $this->assertNull(Validator::getDefaultProvider(null), 'Default provider (null) should be missing');
+
```
Comment (lilHermit) [inline tests/TestCase/Validation/ValidatorTest.php]: Damn it, I'll commit this later with any other changes (if there are)

Comment (dereuromark) [inline tests/TestCase/Validation/ValidatorTest.php]: Adjust your IDE to remove trailing ws for edited lines :)

Comment (lilHermit) [inline tests/TestCase/Validation/ValidatorTest.php]: I normally format the code before hand but the default formatting is so different from your codesniffer rules. Like it wraps long lines but not in the same way as PHPCS wants ;-)

File: src/Validation/Validator.php [hunk 3]
```diff
@@ -53,6 +53,13 @@ class Validator implements ArrayAccess, IteratorAggregate, Countable
     protected $_providers = [];
 
     /**
+     * An associative array of objects or classes used as a default provider list
+     *
+     * @var array
+     */
+    private static $_defaultProviders = [];
```
Comment (dereuromark) [inline src/Validation/Validator.php]: Should be protected. People might have their own extension installed or in use.

File: tests/TestCase/Validation/ValidatorTest.php [hunk 4]
```diff
@@ -1879,4 +1879,36 @@ protected function assertProxyMethod($validator, $method, $extra = null, $pass =
...
     }
+
+    /**
+     * Testing adding DefaultProvider
+     *
+     * @return void
+     */
+    public function testAddingDefaultProvider()
+    {
+        $validator = new Validator;
```
Comment (dereuromark) [inline tests/TestCase/Validation/ValidatorTest.php]: Please add () for class instantiations.

File: tests/TestCase/Validation/ValidatorTest.php [hunk 5]
```diff
@@ -1879,4 +1879,36 @@ protected function assertProxyMethod($validator, $method, $extra = null, $pass =
...
+    }
+
+    /**
+     * Testing getting DefaultProvider(s)
+     *
+     * @return void
+     */
+    public function testGetDefaultProvider()
+    {
+        Validator::addDefaultProvider('test-provider', '\MyNameSpace\Validation\MyProvider');
```
Comment (dereuromark) [inline tests/TestCase/Validation/ValidatorTest.php]: No leading \ needed in string namespaces.

File: src/Validation/Validator.php [hunk 6]
```diff
@@ -211,6 +219,43 @@ public function getProvider($name)
...
+    }
+
+    /**
+     * Associates an object to a name so it can be used as a default provider.
+     *
+     * @param string $name The name under which the provider should be set.
+     * @param object|string $object Provider object or class name.
+     * @return void
+     */
+    public static function addDefaultProvider($name, $object)
```
Comment (ADmad) [inline src/Validation/Validator.php]: Better to name the method `setDefaultProvider()` for consistency.

Comment (dereuromark) [inline src/Validation/Validator.php]: Technically it is adding something to the stack here, setting sounds more like "reset + set".
But I am fine either way.

**Comment by ADmad** [comment]
If your plugin has a behavior you can just use the `Model.buildValidator` event to add provider.

File: src/Validation/Validator.php (hunk 6 above)
Comment (lilHermit) [inline src/Validation/Validator.php]: It was suggested to me by Mark Story (on Slack) to use addDefaultProvider

**Comment by lorenzo** [comment]
how is this supposed to be used?

**Comment by lilHermit** [comment]
@lorenzo 

By adding something like in bootstrap. 

`config/bootstrap.php`
``` php
Validator::addDefaultProvider('extra-validation', 'MyNameSpace\Validation\ExtraValidationProvider');
```

This saves adding the provider in every Table that needs it

**Comment by dereuromark** [comment]
Do we also need a way to clear them, or at least setProviders(array) which would allow an array - and also an emty array to clear?

**Comment by lilHermit** [comment]
@dereuromark Potentially although there is no mechanism for that with setProvider (but it isn't static). Having said that I can't see we where in the lifecycle you would set a defaultProvider and then need to unset, surely if that's the case just use setProvider

File: src/Validation/Validator.php [hunk 7]
```diff
@@ -211,6 +219,43 @@ public function getProvider($name)
     }
 
     /**
+     * Returns the default provider stored under that name if it exists.
+     *
+     * @param string $name The name under which the provider should be retrieved.
+     * @return object|string|null
```
Comment (ravage84) [inline src/Validation/Validator.php]: With three different potential return value types, it might be worth describing which comes in which case.

**Comment by lorenzo** [comment]
Ok, you can add default validators, but where is this used in the rest of the framework?

File: src/Validation/Validator.php (hunk 7 above)
Comment (lilHermit) [inline src/Validation/Validator.php]: I copied the docblock from https://github.com/lilHermit/cakephp/blob/2b580b05c0ee869daf5b38709d37756c5e01f90a/src/Validation/Validator.php#L270 ;-)

**Comment by lilHermit** [comment]
@lorenzo it isn't, it's an enhancement

File: src/Validation/Validator.php (hunk 7 above)
Comment (ravage84) [inline src/Validation/Validator.php]: Not always is existing code, good code 😼

**Comment by ravage84** [comment]
@lilHermit is it intentional that there is no way of removing default providers?

**Comment by lilHermit** [comment]
@ravage84 yeah! Check here https://github.com/cakephp/cakephp/pull/10358#issuecomment-284666420 and its reply

**Comment by dereuromark** [comment]
Will you add a small passage to the docs?

**Comment by lilHermit** [comment]
@dereuromark To the docblock or the actual docs repo?

**Comment by lorenzo** [comment]
@lilHermit that's my point. There is no use for this addition if it is not integrated anywhere in the framework. You may as well just put the provider in `Configure` in that case

**Comment by lilHermit** [comment]
@lorenzo I wasn't aware you could? Do you have an example?

**Comment by lorenzo** [comment]
`Configure::write('whatever', 'value')` This is pretty much how you are using `Validator` now.

What I mean with this is that this change does not make sense to include on its own, it needs to integrate in a way with the rest of the framework. If you are only using the class to store configuration, but the configuration is not used by the framework anywhere, then it should live in `Configure` and not in this class.

**Comment by lilHermit** [comment]
@lorenzo This change isn't just "storing configuration", it is adding the defaultProviders to the providers list in the Validator constructor.

**Comment by lorenzo** [comment]
@lilHermit ah ok, that's the information I was missing. Thanks for clarifying

File: src/Validation/Validator.php (hunk 6 above)
Comment (ADmad) [inline src/Validation/Validator.php]: Okay, makes sense.

**Comment by dereuromark** [comment]
The CakePHP documentation.

**Comment by dereuromark** [comment]
@lorenzo Are you good with the current changes?

**Review by dereuromark — APPROVED** [review]

**Comment by markstory** [comment]
Moving to 3.5.0

**Pull request merged**